### PR TITLE
WIP: Add support for multi-route simulations (recipe books)

### PR DESCRIPTION
**Pull request by SGospodinov**
__This is a WIP PR!__

This PR aims to add support for multi-route simulations. The problem is that such a change will require significant changes in the current REST interface. We propose to replace "recipes" with "recipe books". A recipe book will be an array of JSON object pretty much the same to the recipe objects that the `/add_recipe` endpoint accepts. However, each of the recipes in a recipe book will have an extra `route` attribute. This attribute will represent the endpoint that will return responses for the recipe simulation.

An example recipe book looks like this:

```json
[  
   {  
      "route":"/example/endpoint",
      "body":"Example body",
      "stages":[
         { "at":0, "latency":"400ms", "status":200 },
         { "at":"1s", "latency":"100ms", "status":503}
      ]
   },
   {  
      "route":"/news",
      "origin":"https://www.bbc.co.uk/news",
      "stages":[  
         { "at":0, "status":404, "latency":"50ms" },
         { "at":"2s", "status":503, "latency":"2s" },
         { "at":"4s", "status":200, "latency":"100ms" }
      ]
   }
]
```

Sending a `POST` request to the `/add_recipe_book` endpoint will start new simualtion for each of the recipes in the recipe book. Each of the simulations will be independent of the rest and will be available on the endpoint specified in the `route` attribute.

If we add the above recipe book, the simulation will execute in the following manner
```
/example/endpoint

  0s          1s                                                  ∞
  *───────────*───────────────────────────────────────────────────▶

  HTTP 200 400ms                  HTTP 503 100ms

-------------------------------------------------------------------------------

/news

  0s                     2s                   4s                  ∞
  *──────────────────────*────────────────────*───────────────────▶

       HTTP 404 50ms           HTTP 503 2s       HTTP 200 100ms
```

### CODE REVIEW

**Comment by samfrench** [comment]
Why do we need multiple recipes with multiple stages? Why not have recipes with multiple stages and these recipes could have a route/path they respond to? If not path is provided the default could be "/*"?

Maybe I am missing something in this PR.

The issue we will have though is how they get deleted/replaced if we were to key them on path as the list could increase a lot. Maybe an expiry on each recipe of one day where they would get deleted after a set amount of time.

File: lib/origin_simulator/payload.ex [hunk 1]
```diff
@@ -1,38 +1,34 @@
...
-  defp cache_lookup() do
-    case :ets.lookup(:payload, "body") do
-      [{"body", body}] -> {:ok, body}
+  def restart do
+    GenServer.stop(:payloads)
+  end
+
+  defp cache_lookup(route) do
+    case :ets.lookup(:payloads, route) do
+      [{^route, body}] -> {:ok, body}
```
Comment (ettomatic) [inline lib/origin_simulator/payload.ex]: Currently, our single route acts as a catch-all. 
You are matching by value with the pin operator which will only match exact routes

**Comment by ettomatic** [comment]
Hi @SGospodinov, thank you very much for this, looks promising!

Just a couple of comments:

As Sam said, I really think that we should see the simulation as a single unit of action. The recipe could contain multiple routes but, in my view, it should still be considered as a single _thing_. So, in my opinion, no need to introduce a new concept of a recipe book.

While multiple routes handling is probably necessary (there's also another branch from @james-bowers to try to tackle the same problem) it should still work with a catch-all approach: apart from a few specific route exceptions all the rest of the requests, independently of the path, should be served successfully.
Some possible solutions:
- use `Plug.match`, the complex part here is how to load/reload the recipe as a module, I'd look at what James has done in his branch.
- leave the routes matching to a simple regex. This wouldn't be possible at [ets lookup](https://github.com/bbc/origin_simulator/pull/15/files#diff-1dfa95761da3fe14fc6fdee337ab7530R29) but could pheraps happen just before, by match casing the request path with the list of routes part of the recipe, something like:

```elixir
# simplification of the recipe
routes = %{"/admin/*" => "admin", "/*" => "catch-all"}

routes |> Enum.find(fn {x, _} -> String.match?("/baz", Regex.compile!(x)) end)
# => {"/*", "catch-all"}

routes |> Enum.find(fn {x, _} -> String.match?("/admin/foo/bar", Regex.compile!(x)) end)
# => {"/admin/*", "admin"}
```
then you could use the second item of the returned tuple for the ets lookup..

**Comment by SGospodinov** [comment]
Thanks for the reviews. I didn't realise that the changes will break the catch-all behaviour. I also agree that a regex-based route matching would be useful. I'll try to implement something.

**Comment by ettomatic** [comment]
This can be closed because of https://github.com/bbc/origin_simulator/pull/22
Thanks for all the efforts @SGospodinov - even if we didn't merge it has helped us in shaping the feature.

**Comment by SGospodinov** [comment]
No problem @ettomatic. Working on that was a great way to get introduced to Elixir. Happy to know that the feature is now available. Thanks to the whole team for the time and the effort they put into the implementation. :)

**Pull request closed without merging**